File: src/main/java/brainflow/application/presentation/SearchableImageFileExplorer.java
package brainflow.application.presentation;

import brainflow.gui.AbstractPresenter;
import com.jidesoft.list.ListModelWrapper;
import com.jidesoft.swing.SearchableUtils;
import com.jidesoft.tree.FilterableTreeModel;
import com.jidesoft.tree.QuickTreeFilterField;
import com.jidesoft.plaf.LookAndFeelFactory;
import com.xduke.xswing.DataTipManager;

import javax.swing.*;
import javax.swing.tree.*;
import java.awt.*;

import org.apache.commons.vfs.VFS;

/**
 * Created by IntelliJ IDEA.
 * User: Brad Buchsbaum
 * Date: Dec 14, 2006
 * Time: 7:30:20 PM
 * To change this template use File | Settings | File Templates.
 */
public class SearchableImageFileExplorer extends AbstractPresenter {

    private JPanel mainPanel = new JPanel(new BorderLayout(6, 6));

    private ImageFileExplorer explorer;


    public SearchableImageFileExplorer(ImageFileExplorer explorer) {
        this.explorer = explorer;
        init();
    }

    public static void main(String[] args) {
        try {
            //UIManager.setLookAndFeel(new org.jvnet.substance.skin.SubstanceCremeCoffeeLookAndFeel());
            LookAndFeelFactory.installDefaultLookAndFeel();
            ImageFileExplorer explorer = new ImageFileExplorer(VFS.getManager().resolveFile("C:/javacode"));
            SearchableImageFileExplorer sexplorer = new SearchableImageFileExplorer(explorer);

            JFrame frame = new JFrame();
            frame.add(sexplorer.getComponent(), BorderLayout.CENTER);
            frame.pack();
            frame.setVisible(true);

        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }

    private TreePath getPathForNode(TreeNode current) {
        // Get node depth
        int depth = 0;
        for (TreeNode node = current; node != null; node.getParent())
            depth++;
        // Construct node path
        // First scan helped us, now we can directly allocate array of exact
        // size => no extra objects created (be kind to your local gc, it has
        // hard job cleaning that mess already), no collection reverse.
        // Price is doubling path construction time.
        // But in many situations you know the depth already. In such case
        // Only code below applies and time complexity is O(depth) not
        // O(2*depth)
        TreeNode[] path = new TreeNode[depth];
        for (TreeNode node = current; node != null; node.getParent())
            path[--depth] = node; // reverse fill array
        return new TreePath(path);

    }

    private void init() {
        JPanel quickSearchPanel = new JPanel(new FlowLayout(FlowLayout.LEADING));

        final TreeModel treeModel = explorer.getTreeModel();
        final QuickTreeFilterField field = new QuickTreeFilterField(treeModel) {
            protected FilterableTreeModel createDisplayTreeModel(TreeModel treeModel) {
                return new FilterableTreeModel(treeModel) {
                    @Override
                    protected void configureListModelWrapper(ListModelWrapper wrapper, Object node) {

                        if (node instanceof DefaultMutableTreeNode) {
                            //ImageFileExplorer.LazyNode inode = (ImageFileExplorer.LazyNode) node;
                            DefaultMutableTreeNode inode = (DefaultMutableTreeNode) node;
                            boolean expanded = explorer.getJTree().isExpanded(new TreePath(inode.getPath()));
                            if (inode.isLeaf() || expanded) {
                                super.configureListModelWrapper(wrapper, node);
                            }

                        } 
                    }

                    
                };
            }
        };


        field.setSearchingDelay(200);
        quickSearchPanel.add(field);


        JPanel treePanel = new JPanel(new BorderLayout(2, 2));


        field.setTree(explorer.getJTree());
        SearchableUtils.installSearchable(field.getTree());

        JTree tree = field.getTree();
        tree.setModel(field.getDisplayTreeModel());
        DataTipManager.get().register(tree);


        treePanel.add(explorer.getComponent());

        mainPanel.add(treePanel);
        mainPanel.add(quickSearchPanel, BorderLayout.BEFORE_FIRST_LINE);


    }

    public JComponent getComponent() {
        return mainPanel;
    }

    private String getFilterStatus(TreeModel displayModel, TreeModel originalModel) {
        int count = displayModel.getChildCount(displayModel.getRoot());
        String text = count + " out of " + originalModel.getChildCount(originalModel.getRoot()) + " files";
        return text;
    }

}

File: src/main/scala/sc/bflow/app/presentation/SearchableImageFileExplorer.scala
package sc.bflow.app.presentation

import com.jidesoft.tree. {FilterableTreeModel, QuickTreeFilterField}
import com.jidesoft.list.ListModelWrapper
import javax.swing. {JTree, JPanel}
import com.xduke.xswing.DataTipManager
import com.jidesoft.swing.SearchableUtils
import java.awt. {BorderLayout, FlowLayout}
import org.apache.commons.vfs.FileObject
import boxwood.binding.swing.tree.GenericTreeNode
import javax.swing.tree. {TreeNode, TreePath, DefaultMutableTreeNode, TreeModel}
import collection.mutable.ArrayBuffer
import java.lang.reflect.Field
import java.util.Hashtable

/**
 * Created by IntelliJ IDEA.
 * User: Brad
 * Date: 10/17/10
 * Time: 7:35 PM
 * To change this template use File | Settings | File Templates.
 */

class SearchableImageFileExplorer(roots: FileObject*) extends ImageFileExplorer(roots: _*) {

  val quickSearchPanel: JPanel = new JPanel(new FlowLayout(FlowLayout.LEADING))

  val field: QuickTreeFilterField =

    new QuickTreeFilterField(super.treeModel) {
      protected override def createDisplayTreeModel(treeModel: TreeModel): FilterableTreeModel = {
        new FilterableTreeModel((treeModel)) {
          protected override def configureListModelWrapper(wrapper: ListModelWrapper, node: AnyRef): Unit = {

            if (node.isInstanceOf[GenericTreeNode[_]]) {
              val inode = node.asInstanceOf[GenericTreeNode[_]]
              val ipath = nodePath(inode)
              val expanded: Boolean = treeComponent.isExpanded(ipath)

              if (inode.isLeaf || expanded) {
                super.configureListModelWrapper(wrapper, node)
              }
            }
          }
        }

      }
    }

  def nodePath(_node: TreeNode) = {
    var node = _node
    val list = ArrayBuffer[AnyRef]()
    while (node != null) {
      list += node
      node = node.getParent();
    }

    new TreePath(list.reverse.toArray)


  }


    field.setSearchingDelay(200)
    field.setTree(treeComponent)

    quickSearchPanel.add(field)
    println(field.getTree)
    SearchableUtils.installSearchable(field.getTree)



    field.getTree.setModel(field.getDisplayTreeModel)
    DataTipManager.get.register(field.getTree)
    add(quickSearchPanel, BorderLayout.BEFORE_FIRST_LINE)


}